[ONNX] Add `acceptable_error_percentage` to backend tests

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82623
* __->__ #82622
* #82621

To enable more test coverage for numerically unstable tests that
occasionally fails with sporadic element mismatch. Currently these
tests have to be disabled in CI. This PR provides a configurable
tolerance threshold for such issues.

BowenBao committed:
[ONNX] Add `acceptable_error_percentage` to backend tests

To enable more test coverage for numerically unstable tests that
occasionally fails with sporadic element mismatch. Currently these
tests have to be disabled in CI. This PR provides a configurable
tolerance threshold for such issues.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Add `acceptable_error_percentage` to backend tests"


To enable more test coverage for numerically unstable tests that
occasionally fails with sporadic element mismatch. Currently these
tests have to be disabled in CI. This PR provides a configurable
tolerance threshold for such issues.

[ghstack-poisoned]

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -113,14 +113,29 @@ def _ort_session(
 
     assert len(pt_outs) == len(ort_outs), "number of outputs differ"
+    if acceptable_error_percentage:
+        assert (
```
I would raise a ValueError for user mistakes.

justinchuby commented on test/onnx/test_models.py:
```diff
@@ -143,13 +143,11 @@ def test_resnet(self):
-        "This test has been flaky on trunk and PRs. See https://github.com/pytorch/pytorch/issues/79540"
-    )
     @skipScriptTest(min_opset_version=15)  # None type in outputs
+    # This test is numerically unstable. Sporadic single element mismatch occurs occasionally.
```
Trying to understand context: Is single element mismatch not a bug?

BowenBao replied:
Good question. I think the backend folks would have a better say. It's definitely in a gray area. But I wouldn't want to disable the entire test until everything is fixed either. 
rtol/atol won't work in this case. It's a kind of "spike" mismatch, very large diff in value.

justinchuby replied:
Gotcha!

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -113,14 +113,29 @@ def _ort_session(
-def _compare_ort_pytorch_outputs(ort_outs, pt_outs, rtol, atol):
+def _compare_ort_pytorch_outputs(
+    ort_outs, pt_outs, rtol, atol, acceptable_error_percentage: Optional[float]
+):
```
It'd be helpful to define and explain `acceptable_error_percentage` in the docstring here.

BowenBao replied:
done

BowenBao committed:
Update on "[ONNX] Add `acceptable_error_percentage` to backend tests"


To enable more test coverage for numerically unstable tests that
occasionally fails with sporadic element mismatch. Currently these
tests have to be disabled in CI. This PR provides a configurable
tolerance threshold for such issues.

[ghstack-poisoned]

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -116,14 +116,34 @@ def _ort_session(
+    if (
+        acceptable_error_percentage
+        and acceptable_error_percentage <= 1.0
+        and acceptable_error_percentage >= 0.0
```
Should the logic be reversed?

BowenBao committed:
Update on "[ONNX] Add `acceptable_error_percentage` to backend tests"


To enable more test coverage for numerically unstable tests that
occasionally fails with sporadic element mismatch. Currently these
tests have to be disabled in CI. This PR provides a configurable
tolerance threshold for such issues.

[ghstack-poisoned]

BowenBao commented:
@abock for merge approval

BowenBao committed:
Update on "[ONNX] Add `acceptable_error_percentage` to backend tests"


To enable more test coverage for numerically unstable tests that
occasionally fails with sporadic element mismatch. Currently these
tests have to be disabled in CI. This PR provides a configurable
tolerance threshold for such issues.

[ghstack-poisoned]

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -147,7 +155,18 @@ def _compare_ort_pytorch_outputs(
         try:
-            np.testing.assert_allclose(ort_out, pt_out, rtol=rtol, atol=atol)
+            # TODO: Remove `check_shape` option once every shape inconsistent issue is addressed.
+            if not check_shape:
```
Indent?

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -122,27 +122,59 @@ def _compare_ort_pytorch_outputs(
+                    np.isclose(ort_out, pt_out, rtol=rtol, atol=atol)
+                ) / np.prod(ort_out.shape)
+                if error_percentage <= acceptable_error_percentage:
+                    continue
```
I think it would be helpful to produce a warning here

abock replied:
We should definitely log a warning and probably coalesce similar warnings into a summary warning if log pollution is an issue.

BowenBao replied:
Agreed. I will be adding a warning for now.

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -122,27 +122,59 @@ def _compare_ort_pytorch_outputs(
+                ) / np.prod(ort_out.shape)
+                if error_percentage <= acceptable_error_percentage:
+                    continue
+            raise e
```
```suggestion
            raise
```

abock reviewed (changes requested):
@justinchuby has good feedback here

abock reviewed (changes requested):
@justinchuby has good feedback here

BowenBao committed:
Update on "[ONNX] Add `acceptable_error_percentage` to backend tests"


To enable more test coverage for numerically unstable tests that
occasionally fails with sporadic element mismatch. Currently these
tests have to be disabled in CI. This PR provides a configurable
tolerance threshold for such issues.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g
